File: src/tickoni/runtime/sandbox.zig
/// v2.14.S8.T5: this struct is defined and tested, but not yet wired into
/// the process-mode harness call path (c_abi/shim/tile_run.c's
/// tk_topo_run_tile_simple keeps sandbox=0 throughout v2.14). Real entry
/// via c_abi.sandbox.enter needs unshare(CLONE_NEWUSER) to drop to this
/// struct's default nobody uid/gid, which is not guaranteed to be
/// available — confirmed by a concrete EPERM failure under this
/// implementation's own sandboxed environment despite the kernel allowing
/// unprivileged user namespaces (host LSM/AppArmor policy can still deny
/// it). Requiring that privilege in the harness path would make Linux
/// full-runtime process mode depend on something consumer-tier hosts
/// (V1.21) and hardened hosts may not have — process isolation here comes
/// from supervisor-managed OS processes and crash-only teardown, not
/// namespace confinement, so v2.14 does not need this to hold. Wiring
/// real sandbox entry (including a documented fail-closed-with-diagnostic
/// path when namespaces are unavailable, never escalating privileges) is
/// scoped to a future V1.21 support-tier story. See
/// doc/strategy/roadmap/stories/v2.14.md's v2.14.S8.T5 entry.
const builtin = @import("builtin");
const std = @import("std");
const util = @import("util");
const sandbox_defaults = util.sandbox_defaults;

/// Process-mode sandbox policy tier. v2.14 keeps the retail/consumer path on
/// explicit `none` across every platform: no sudo, no capabilities, and no
/// namespace dependency. A future Linux-only support tier can add a real
/// namespace sandbox mode without changing macOS semantics.
pub const SandboxMode = enum {
    none,
    linux_namespaces,
};

pub fn defaultProcessModeForOs(os_tag: std.Target.Os.Tag) SandboxMode {
    _ = os_tag;
    return .none;
}

pub fn defaultProcessMode() SandboxMode {
    return defaultProcessModeForOs(builtin.os.tag);
}

pub fn processModeUsesSandbox(mode: SandboxMode) bool {
    return mode != .none;
}

/// Configuration for entering a tile sandbox via c_abi.sandbox.enter.
/// Defaults are the most restrictive safe values.
pub const SandboxConfig = struct {
    desired_uid: u32 = sandbox_defaults.nobody_uid,
    desired_gid: u32 = sandbox_defaults.nobody_gid,
    keep_host_networking: bool = false,
    allow_connect: bool = false,
    allow_renameat: bool = false,
    keep_controlling_terminal: bool = false,
    dumpable: bool = false,
    rlimit_file_cnt: u64 = sandbox_defaults.rlimit_file_cnt,
    rlimit_address_space: u64 = sandbox_defaults.rlimit_address_space,
    rlimit_data: u64 = sandbox_defaults.rlimit_data,
    rlimit_nproc: u64 = 0,
};

test "default process mode is explicit none on this build" {
    try std.testing.expectEqual(SandboxMode.none, defaultProcessMode());
    try std.testing.expect(!processModeUsesSandbox(defaultProcessMode()));
}

test "macOS process mode remains none" {
    try std.testing.expectEqual(SandboxMode.none, defaultProcessModeForOs(.macos));
    try std.testing.expect(!processModeUsesSandbox(defaultProcessModeForOs(.macos)));
}

test "linux process mode stays none until a future support-tier story wires namespaces" {
    try std.testing.expectEqual(SandboxMode.none, defaultProcessModeForOs(.linux));
    try std.testing.expect(!processModeUsesSandbox(defaultProcessModeForOs(.linux)));
}

test "SandboxConfig defaults are least-privilege" {
    const cfg = SandboxConfig{};
    try std.testing.expectEqual(sandbox_defaults.nobody_uid, cfg.desired_uid);
    try std.testing.expectEqual(sandbox_defaults.nobody_gid, cfg.desired_gid);
    try std.testing.expect(!cfg.keep_host_networking);
    try std.testing.expect(!cfg.allow_connect);
    try std.testing.expect(!cfg.allow_renameat);
    try std.testing.expect(!cfg.dumpable);
    try std.testing.expectEqual(sandbox_defaults.rlimit_file_cnt, cfg.rlimit_file_cnt);
    try std.testing.expectEqual(sandbox_defaults.rlimit_data, cfg.rlimit_data);
    try std.testing.expectEqual(@as(u64, 0), cfg.rlimit_nproc);
}

test "SandboxConfig rlimit_address_space is at least the documented minimum" {
    const cfg = SandboxConfig{};
    try std.testing.expectEqual(sandbox_defaults.rlimit_address_space, cfg.rlimit_address_space);
    try std.testing.expect(cfg.rlimit_address_space >= sandbox_defaults.min_rlimit_address_space);
}

File: src/tickoni/c_abi/shim/sandbox.c
/* Thin wrappers around Firedancer sandbox primitives. */

#include "../../../util/fd_util.h"

/* Process-mode sandbox policy: Linux may use real Firedancer sandbox entry in a
   future support tier, but the current retail/consumer path keeps macOS,
   Windows, and every other non-Linux target on explicit sandbox=none: no
   sudo, no capabilities, and no namespace dependency. */

/* Sandbox functions are Linux-only (seccomp, user namespaces). */
#if FD_HAS_LINUX

#include "fd_sandbox.h"

int tk_sandbox_requires_cap_sys_admin( uint desired_uid, uint desired_gid ) { return fd_sandbox_requires_cap_sys_admin( desired_uid, desired_gid ); }

void
tk_sandbox_enter( uint        desired_uid,
                  uint        desired_gid,
                  int         keep_host_networking,
                  int         allow_connect,
                  int         allow_renameat,
                  int         keep_controlling_terminal,
                  int         dumpable,
                  ulong       rlimit_file_cnt,
                  ulong       rlimit_address_space,
                  ulong       rlimit_data,
                  ulong       rlimit_nproc,
                  ulong       allowed_file_descriptor_cnt,
                  int const * allowed_file_descriptor,
                  ulong       seccomp_filter_cnt,
                  void *      seccomp_filter ) {
  fd_sandbox_enter( desired_uid,
                    desired_gid,
                    keep_host_networking,
                    allow_connect,
                    allow_renameat,
                    keep_controlling_terminal,
                    dumpable,
                    rlimit_file_cnt,
                    rlimit_address_space,
                    rlimit_data,
                    rlimit_nproc,
                    allowed_file_descriptor_cnt,
                    allowed_file_descriptor,
                    seccomp_filter_cnt,
                    seccomp_filter );
}

void tk_sandbox_switch_uid_gid( uint desired_uid, uint desired_gid ) { fd_sandbox_switch_uid_gid( desired_uid, desired_gid ); }
int tk_sandbox_getpid( void ) { return fd_sandbox_getpid(); }
int tk_sandbox_gettid( void ) { return fd_sandbox_gettid(); }

#elif FD_HAS_WINDOWS

#include <windows.h>

int tk_sandbox_requires_cap_sys_admin( uint desired_uid, uint desired_gid ) { (void)desired_uid; (void)desired_gid; return 0; }
void tk_sandbox_enter( uint desired_uid, uint desired_gid, int keep_host_networking, int allow_connect, int allow_renameat, int keep_controlling_terminal, int dumpable, ulong rlimit_file_cnt, ulong rlimit_address_space, ulong rlimit_data, ulong rlimit_nproc, ulong allowed_file_descriptor_cnt, int const * allowed_file_descriptor, ulong seccomp_filter_cnt, void * seccomp_filter ) {
  (void)desired_uid; (void)desired_gid; (void)keep_host_networking; (void)allow_connect; (void)allow_renameat; (void)keep_controlling_terminal; (void)dumpable; (void)rlimit_file_cnt; (void)rlimit_address_space; (void)rlimit_data; (void)rlimit_nproc; (void)allowed_file_descriptor_cnt; (void)allowed_file_descriptor; (void)seccomp_filter_cnt; (void)seccomp_filter;
}
void tk_sandbox_switch_uid_gid( uint desired_uid, uint desired_gid ) { (void)desired_uid; (void)desired_gid; }
int tk_sandbox_getpid( void ) { return (int)GetCurrentProcessId(); }
int tk_sandbox_gettid( void ) { return (int)GetCurrentThreadId(); }

#else

/* Hosted retail stubs — sandbox not supported on this platform. */
#include <unistd.h>

int tk_sandbox_requires_cap_sys_admin( uint desired_uid, uint desired_gid ) { (void)desired_uid; (void)desired_gid; return 0; }
void tk_sandbox_enter( uint desired_uid, uint desired_gid, int keep_host_networking, int allow_connect, int allow_renameat, int keep_controlling_terminal, int dumpable, ulong rlimit_file_cnt, ulong rlimit_address_space, ulong rlimit_data, ulong rlimit_nproc, ulong allowed_file_descriptor_cnt, int const * allowed_file_descriptor, ulong seccomp_filter_cnt, void * seccomp_filter ) {
  (void)desired_uid; (void)desired_gid; (void)keep_host_networking; (void)allow_connect; (void)allow_renameat; (void)keep_controlling_terminal; (void)dumpable; (void)rlimit_file_cnt; (void)rlimit_address_space; (void)rlimit_data; (void)rlimit_nproc; (void)allowed_file_descriptor_cnt; (void)allowed_file_descriptor; (void)seccomp_filter_cnt; (void)seccomp_filter;
}
void tk_sandbox_switch_uid_gid( uint desired_uid, uint desired_gid ) { (void)desired_uid; (void)desired_gid; }
int tk_sandbox_getpid( void ) { return getpid(); }
int tk_sandbox_gettid( void ) { return getpid(); } /* gettid() is Linux-only; use getpid() as fallback */

#endif /* FD_HAS_LINUX */
